src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,153 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Analyzer.Utilities;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+
+namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
+{
+    // TODO: VisualBasic as well?
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
+    public sealed class StringBuilderAppendShouldNotTakeSubstringFixer : CodeFixProvider
+    {
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray
+            .Create(
+                StringBuilderAppendShouldNotTakeSubstring.RuleIdOneParameterId,
+                StringBuilderAppendShouldNotTakeSubstring.RuleIdTwoParameterId);
+
+        public sealed override FixAllProvider GetFixAllProvider()
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            SyntaxNode root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            SyntaxNode node = root.FindNode(context.Span);
+
+            // TODO is it guaranteed that there's only one at once?
+            Diagnostic diagnostic = context.Diagnostics.Single();
+            if (diagnostic.Id == StringBuilderAppendShouldNotTakeSubstring.RuleIdOneParameterId)
+            {
+                context
+                    .RegisterCodeFix(
+                        new MyCodeAction(
+                            MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithOneParameterFix,
+                            async ctx => await FixCodeOneParameter(context.Document, root, node, ctx).ConfigureAwait(false),
+                            equivalenceKey: MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithOneParameterFix),
+                        diagnostic);
+            }
+            else
+            {
+                context
+                    .RegisterCodeFix(
+                        new MyCodeAction(
+                            MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithTwoParameterFix,
+                            async ctx => await FixCodeTwoParameters(context.Document, root, node, ctx).ConfigureAwait(false),
```

Comment:
in general, it's not advised to capture things like root/node here.  These lambdas will then keep that data alive and prevent the IDE from doing things like dumping the in-memory trees.  That said, the chance of everyone doing the right thing and not capturing is effectively nil at this point... so i'm on hte fence about changing these things :(

Reply:
Changed, now only using the input context and the cancellationToken.